### DISCUSSION TITLE
Permissions Added

### DISCUSSION THREAD

**Discussion by NathanaelA** [NativeScript/rfcs#10] [1. Ideas & Discussions]
I think NS 8.0 should have the NS-Permissions plugin built in.   The Permissions plugin is the most widely used plugin in the entire eco-system, and even used by some of the NS standard plugins.    Seems to make sense to integrate it into the core so that it is no longer a pretty much required "add-on".

I don't know if we should work on making the API a bit more generic to handle IOS as the current plugin is Android only, but it a thought to allow both the current method and adding helper features to support IOS & Android permissions like 
Permission.get(CAMERA).then(...) where it asks for both Android and IOS permissions at the same time?

**Comment by farfromrefug** [comment]
I actually dont think it should be in the core. I think N should be splitted as much as possible. I dont see any good reason to put it "in". Just like I think the ui should be splitted into a separated plugin. 
Now that being said I would go for the community permissions plugin. " @nativescript-community/perms" already supports iOS and Android. It is also fully written in typescript l. It is a direct port of react native permissions plugin, which has the advantage of making it easier for users to switch to N in a sense. 
I agree there could/should be a first class permissions plugin in N.

**Comment by NathanaelA** [comment]
Your statement is very confusing. 
>I actually dont think it should be in the core. 
>I agree there could/should be a first class permissions plugin in N.

You state you don't want permissions in core but then on the other hand you do want permissions in core?   

---

Well, let see, which does the community actually use should probably determine the base design: 
- NS-Permissions: 5,808 per week (NPM) + ~1,000 per week (ProPlugins, same plugin different namespace)
- Community/NS-Perms : 183 per week (for Community) & 74 per week (for prior non-community repo version).    

---
>it is also fully written in typescript 

And the plugin would obviously have to be converted to TS as the entire base of NS is TS we aren't going to start mixing code, as much as I prefer to use ES6 JS myself, I totally believe mixing the NS code base is 100% the wrong direction, so that means any future code we put in the NS-Core should 100% TS.  :grinning: 

--- 
  
I think we probably should have a vote on direction; as I'm obviously biased and pro the code I wrote and the way I wrote it, and Martin is obviously biased and pro the way he wrote his code.    :grinning:

A couple possible ways to go:
1. :heart: - Use the Community/ns-perms and beef up the support as it is missing some of cross platform items it really should have and also fix the memory leaks.

2. :+1: - Use my plugin but enhance it to support the iOS permissions using something like `Permission.Camera.request()` So it has a class specific to each of the iOS permission (about 10 iOS Specific permissions) each one should be a child name class, so it is easy to use and find for maintenance.     Also create the same classes on the Android side for the same permissions, so that either platform can do `Camera.Request()` and both sides would do the proper thing...   Import would be `import {Camera} from ...` so it is easy to use... 

Please note I'm not 100% set on the cross platform API design, but I feel it should be abstracted out class groups to be easy to upgrade as OS upgrades occur that require additional permissions to be requested for what might have been a single permission, so being able to easily find the `Permissions.Location.request` and add the additional checks for API 30 that weren't required (or even existed) in API 19...

**Comment by farfromrefug** [comment]
Just One thing please @NathanaelA stop with your download number thing you keep on bringing it up everywhere. It makes no sense. Of course a plugin  there for so much more time will have bigger number. 
With that kind of argumentation you re like going against progress / change...
Now my point was having a good cross platform perms in @nativescript scope makes sense.
Just dont put it in the core !!! 
For one it gives people the choice to use one or another plugin.
I dont really care which one you decide to go for. If I dont like it I can simply keep on using the other ;)

And we might want to keep one in the community. Much eqisier for us to maintain and release as we use it many community plugins already. We dont have to wait for N to release a fix or something when we need a change we can simply release.

So choose as you prefer I ll support whatever the choice is as long as it stays outside of N ;)

**Comment by NathanaelA** [comment]
>Just One thing please @NathanaelA stop with your download number thing you keep on bringing it up everywhere.

I bring up the download number to point out what is being used heavily and what is actually _battle_ tested.   I will always prefer using actual evidence and not just feelings.  So no, I will always attempt to use real data and that may include download numbers, usage, speed, performance, tests, etc.   Real data always trumps thoughts..


>  It makes no sense. Of course a plugin there for so much more time will have bigger number.

Yes, I agree my plugin has been out ~3 years longer than yours, but yours still has been out almost 2 years now, so it has had _plenty_ of time to get a following and for people to start using it as we get new people in the eco-system all the time.   However of the 8 dependencies using the ns-perms plugin, 5 are your own.    Of the 74 dependencies (not including any ProPlugins, just 74 on NPMJS) using NS-Permissions, I think 3 are mine. :grinning:        I totally agree some of that is the age of the plugin, but some of that is just because people use what works and what is solid... You are a great developer, and you have contributed some great stuff, but ns-perms is not a high quality work, and as such I would tend to think some of the numbers using my plugin are because others saw the same issues when looking at it, that I saw when I quickly scanned it the other day while making this RFC. 

> And we might want to keep one in the community. Much eqisier for us to maintain and release as we use it many community plugins already. We dont have to wait for N to release a fix or something when we need a change we can simply release.

The code is already rock solid and battle tested, in the last _2_ years the only changes I've had to do is added the AndroidX support, and change the require from tns-core modules to core.      I also added a couple special permissions that the community asked for that of course Google required a completely different permission work flow.      I have no issues putting my code right now as-is in core because it doesn't need any maintenance, the last time I touched the plugin repo was to upgrade the NS7 demo...   So leaving it outside of core because of bug fixes is not a valid concern at this point...

---

Please note leaving NS-Permissions outside of Core is significantly better for **_ME_** personally as I can point to it to clients and customers showing how heavily and trusted my code is in the entire NativeScript eco-system when they ask.    NS-Permissions really is my most widely used plugin.   So I'm trading street cred for invisibility as I already have plenty of my code in NS over the years, so adding some more code does nothing for my street creds.    

However, I think to promote growth in the NativeScript framework we have to make the learning ramp smaller, by adding certain features to core that pretty much every app needs.   One of the features that I personally believe has been missing since the beginning has been permissions, and even the prior team had integrating Permissions on their roadmap..    

However, I think we are going off topic with this.   So redirecting...   
---

First question that needs to be decided:
:+1:  or :-1: Should we add some sort of permissions support into Core so it ships with NativeScript and all plugins can start using it by default and can use the exact same interface?


Please click the :+1: or :-1: below to approve or disapprove this -- lets just see what the community prefers before we continue on with discussion.

**Comment by MrSnoozles** [comment]
My two cents: 
_(I might have too little knowledge about internals though for a proper judgement.)_

From what you are communicating you're planning to extend the scope of NativeScript, e.g. with #11 . So while more platforms are being added the NS team would have to make sure the core permissions work for every platform. This will result in additional development and maintainance efforts. 
Nathanael, you wrote that there was almost no maintainance required for your permissions plugin. This might change though as more platforms are being targeted.

If permissions were available as a plugin we could simply use your plugin for iOS and Android and use another one for another platform without it feeling "wrong". 


---

> Seems to make sense to integrate it into the core so that it is no longer a pretty much required "add-on".

Additionally, there are a few apps that don't require any sort of permission management. Don't know if integrated core permissions would be disadvantageous in any form for those apps.

**Comment by NathanaelA** [comment]
@MrSnoozles - That is a good thinking, thanks!        However, if we were to add #11 and used QODE, then there actually is no additional permissions needed as QODE is Node with QT embedded.      So no additional resources needed or maintenance.   Also desktop typically does NOT need any additional permission support, most desktop OS's assume you have all rights except root/admin.   And very little is blocked off from a typically running desktop app.     However, I do appreciate the time you took to think about this possibility, but I believe that is a non-issue at this point.

---

I would honestly be surprised if their are very many apps in the eco-system that don't use my permission plugin already as a dependency.    
If you use any plugin that talks to a camera, sensor, bluetooth, contacts, or any number of 100's of other things that requires permissions, it is probably already using my plugin.   In fact their are several plugins core maintained plugins (i.e. `@nativescript/PLUGINS`) that are already using my plugin because the prior core team chose to use mine rather than re-implement it.     

Just for fun, you should scan your `node_modules` and see if `nativescript-permissions` or `@proplugins/nativescript-permissions` shows up and how many times...   ;-)   

**However, it was on the roadmap to integrate permissions into the framework, probably because of the following two reasons:**

1. Adding it to core would just mean that they wouldn't get the chance of conflicting versions; if Plugin A requires v1.0 and Plugin B requires v1.1 then npm will give you two copies of the plugin.  This means your code base in your app is inflated by the cost of the plugin a second time.   Now multiply that by 10 plugins that require permission and you could easily have 10 or more copies of NS-Permissions in your app and not even know it.     With multiple versions of the plugin, you've just inflated your app a lot for NO reason.   If NS-Permissions is added to core; the only way you should get it into your awesome app that doesn't need any permissions is if you erroneously add a `import` to it; otherwise webpack shouldn't even pull it in, meaning that apps that don't need it don't get it and apps that do need it only get ONE version of it.   

2. Most apps require permissions of some sort, why make a new user to the eco-system go looking for a permission plugin, which permission plugin do you choose?    Who do you trust?    That is like asking a new user to go out and look for other Layouts, because we don't ship any layouts except DockLayout.    I'm sure their are some apps that only use a DockLayouts, but it makes it a lot easier for a new user to have StackLayout and GridLayout as part of the core because those are commonly used in most apps.
